File: 1.HTTP-AND-CORS/code/typescript/status_codes_7.ts
import express, { Request, Response } from 'express';

const app = express();

const db = {
    find: async (id: string) => {
        if (id !== "1") throw new Error("ErrNotFound");
        return { id, visibleTo: (req: Request) => req.header('Authorization') === 'user' };
    }
};

async function getUser(req: Request, res: Response) {
    if (!req.header('Authorization')) {
        res.status(401).send('login required'); // 401: who are you?
        return;
    }
    
    try {
        const user = await db.find(req.params.id);
        if (!user.visibleTo(req)) {
            res.status(403).send('forbidden'); // 403: not allowed
            return;
        }
        res.status(200).json(user); // 200
    } catch (err: any) {
        if (err.message === "ErrNotFound") {
            res.status(404).send('no such user'); // 404
            return;
        }
        res.status(500).send('Server Error');
    }
}

File: 1.HTTP-AND-CORS/code/rust/status_codes_7.rs
use actix_web::{web, HttpRequest, HttpResponse, Responder};
use serde::Serialize;

#[derive(Serialize)]
struct User {
    id: String,
}

impl User {
    fn visible_to(&self, req: &HttpRequest) -> bool {
        if let Some(auth) = req.headers().get("Authorization") {
            return auth.to_str().unwrap_or("") == "user";
        }
        false
    }
}

async fn get_user(req: HttpRequest, id: web::Path<String>) -> impl Responder {
    if req.headers().get("Authorization").is_none() {
        return HttpResponse::Unauthorized().body("login required"); // 401: who are you?
    }
    
    let id_val = id.into_inner();
    
    // simulated db.Find
    if id_val != "1" {
        return HttpResponse::NotFound().body("no such user");       // 404
    }
    
    let user = User { id: id_val };
    
    if !user.visible_to(&req) {
        return HttpResponse::Forbidden().body("forbidden");         // 403: not allowed
    }
    
    HttpResponse::Ok().json(user)                                   // 200
}
